### DISCUSSION TITLE
Java S3 client cannot access files after updating to 1.4.0

### DISCUSSION THREAD

**Discussion by jjimenezgarcia** [apache/ozone#6650] [FAQ] [answered]
Hello, I'm having an issue after an update from 1.3.0 to 1.4.0.

I access my files through s3 client Java API. My buckets keep their layout properly after updating, so I assume that is not the problem.
After the update, I can no more access those files from my Java application (but I can through command line, so files do exist, they have not disappeared). The strangest is that I can upload files by the same Java application, and only then I can access those new files. Which means, I lose access to all pre-existing files, but can work with new ones without problem.

Hope you can help. This is my Java code:

![imagen](https://github.com/apache/ozone/assets/99907380/cafa4200-5be8-41a7-a753-c9d6f4f74ee9)

**Comment by ChenSammi** [comment]
Are there any logs or error messages can be shared when access fails?

**Reply by jjimenezgarcia** [reply]
Just this:
```bash
2024-05-08T09:31:47.707Z ERROR 1 --- [nio-80-exec-159] o.a.c.c.C.[.[.[/].[dispatcherServlet]    : Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed: java.lang.ArrayIndexOutOfBoundsException] with root cause

java.lang.ArrayIndexOutOfBoundsException: null
```

In addition, it is retrocompatible, files added on 1.4.0 can be properly accessed on both 1.4.0 and 1.3.0, but not the other way around.

**Reply by ivandika3** [reply]
@jjimenezgarcia It maybe a bug related to https://github.com/apache/ozone/pull/6377 which was found on 1.4.0 release and planned for 1.4.1 release. 

Just verifying, was the log you provided thrown from client-side (S3 client)? Could you help to check whether there were related logs on S3G (or if possible OM as well)?

**Reply by jjimenezgarcia** [reply]
The log provided was thrown from the client-side, yes. 

And, as far as I can see, there are no related logs on both S3G or OM.

**Reply by ivandika3** [reply]
@jjimenezgarcia Is it possible to backport the fix https://github.com/apache/ozone/pull/6377 into the current 1.4.0 version and redeploy the S3G with the updated version and let us know whether the issue is resolved?

**Reply by jjimenezgarcia** [reply]
Don't know if that's the issue. Cause all ETags seem to be properly initialized. I attach you both examples of files, one was added on 1.4.0 (first one) and can be accessed, and the other one on 1.3.0 (can not be accessed).
![imagen](https://github.com/apache/ozone/assets/99907380/89a6aaaa-3991-4d56-a67e-d723eb27e107)
![imagen](https://github.com/apache/ozone/assets/99907380/14d49d9c-393d-40ff-aa04-c73dcf3c0459)

I am both uploading and downloading using S3.

**Reply by ivandika3** [reply]
@jjimenezgarcia Thank you for the example. From the example, the key name looks the same. Was it overwritten? And in what order (looks like key that's created on 1.4.0 was created first and then was overwritten by key created on 1.3.0)? Could you help try whether overwriting keys have anything to do with the bug?

One thing I found odd from the example is that the ETag of key created on 1.4.0 should have hash-like structure (`6b3f1d2b18292bbb32585cf0ff31b36c`) instead of date-like structure (as shown from the example).

Another question to help debugging: Was the key uploaded normally or using multipart upload? From the size, it should be normal upload, but just to make sure.

**Reply by ivandika3** [reply]
Also may I ask whether the exception was thrown only when downloading through AWS Java SDK (using the code)? Or does AWS S3 CLI (`aws s3` or `aws s3api`) threw the same error as well?

**Reply by jjimenezgarcia** [reply]
@ivandika3 the key is not overwritten, It's just an internal naming, keys are cut on the screenshot, but longer and different at the end.

ETag has always been a date-like structure, don't ask me why but that happens in every file.

And the key was uploaded as an inputStream. This way:

![imagen](https://github.com/apache/ozone/assets/99907380/c066af95-d117-4721-936d-6e2a6ffbf8e1)

The error is only thrown when downloading trough Java (using that code). Both from command line (aws s3api)  and python, files can be downloaded without any problem.

**Reply by adoroszlai** [reply]
Date-like ETag used to be added in get bucket response:

https://github.com/apache/ozone/blob/e4b6007c8e9f79d1432cf6f128d6f83e733f8e9b/hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java#L717

Fixed in HDDS-10615.

**Reply by jjimenezgarcia** [reply]
Can read "fix version: 1.5.0"

 Is that version even out?

**Reply by adoroszlai** [reply]
1.5.0 is not released yet.

**Reply by jjimenezgarcia** [reply]
But don't know if the ETag is the only responsible. As I told @ivandika3 files can be accessed without problem using 'aws s3api' from command line or python. 

Java seems to be my only problem.

**Reply by adoroszlai** [reply]
> files can be accessed without problem using 'aws s3api' from command line or python.

The same situation was described HDDS-10521 which @ivandika3 pointed out.

Can you please check `ETag` in `aws s3api get-object` response?

**Reply by ivandika3** [reply, marked as answer]
Could you try to set `System.setProperty("com.amazonaws.services.s3.disableGetObjectMD5Validation", "true");` before initializing the S3 Java client and see whether the file can be downloaded?

Ref: https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/internal/SkipMd5CheckStrategy.html

The reason I suspected the Etag, was because the issue only comes up when downloading using AWS Java SDK, although it worked on AWS CLI.

You can also try to debug it in the IDE and let us know which part of code was causing the issue.

**Reply by jjimenezgarcia** [reply]
You are a 100% right. I was just listing the objects and getting those dated `ETag`. Sorry for the prior misunderstanding.

When I download it with get-object. Old files (uploadded on 1.3.0) have no `ETag`, as shown here:

``` java
{
    "AcceptRanges": "bytes",
    "LastModified": "Wed, 08 May 2024 09:26:32 GMT",
    "ContentLength": 31503,
    "ETag": "\"null\"",
    "CacheControl": "no-cache",
    "ContentType": "application/octet-stream",
    "Expires": "Thu, 09 May 2024 07:44:06 GMT",
    "Metadata": {}
}

```

But new files do have a hashed `ETag`, as shown here in a 1.4.0 added file:

``` java
{
    "AcceptRanges": "bytes",
    "LastModified": "Tue, 07 May 2024 11:46:40 GMT",
    "ContentLength": 1461,
    "ETag": "\"6838af242c1394bc088d0d9991256a29\"",
    "CacheControl": "no-cache",
    "ContentType": "application/octet-stream",
    "Expires": "Thu, 09 May 2024 07:44:51 GMT",
    "Metadata": {}
}
```
So that's the reason of the `ArrayIndexOutOfBoundsException`.

What would the solution be? I'm deploying `S3G` with docker image `apache/ozone:1.4.0`, which I see is the latest docker image released.

EDIT: I will try your solution @ivandika3, answered while you were writting, dind't see your message.

**Reply by jjimenezgarcia** [reply]
Sorry  @ivandika3 , do you know how I could initialize that inside the docker container?

Trying to do so but clearly stuck, as the app gets initialized not with a jar file, but with the specific `ozone` command inside the container.

**Reply by ivandika3** [reply]
@jjimenezgarcia Where does your AWS S3 client code deployed? It's a client-side property, just need to include it in the client-side code. There is no need to redeploy anything server-side.

**Reply by jjimenezgarcia** [reply]
Oh sorry, I thought it was a server-side configuration. Will try then.

**Reply by jjimenezgarcia** [reply]
Thank you both @ivandika3 @adoroszlai, the problem got fixed with the property `System.setProperty("com.amazonaws.services.s3.disableGetObjectMD5Validation", "true");` set in the client.

As I understood, this will be unnecessary from version 1.4.1, so will keep this solution as temporary till then.

Thanks!
